File: src/pcl-bindings/VectorFloat.cpp
#include "export.h"
#include <vector>

using point_t = float;
using vector_t = std::vector<point_t>;

EXPORT(vector_t*) std_vector_float_ctor()
{
    return new vector_t();
}

EXPORT(vector_t*) std_vector_float_ctor_count(size_t count)
{
    return new vector_t(count);
}

EXPORT(void)std_vector_float_delete(vector_t** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

EXPORT(void) std_vector_float_at(vector_t* ptr, size_t idx, point_t* value)
{
    *value = ptr->at(idx);
}

EXPORT(size_t) std_vector_float_size(vector_t* ptr)
{
    return ptr->size();
}

EXPORT(void) std_vector_float_clear(vector_t* ptr)
{
    ptr->clear();
}

EXPORT(void) std_vector_float_resize(vector_t* ptr, size_t size)
{
    ptr->resize(size);
}

EXPORT(void) std_vector_float_add(vector_t* ptr, point_t value)
{
    ptr->push_back(value);
}

EXPORT(void) std_vector_float_insert(vector_t* ptr, ptrdiff_t index, point_t value)
{
    ptr->insert(ptr->begin() + index, value);
}

EXPORT(point_t*) std_vector_float_data(vector_t* ptr)
{
    return ptr->data();
}

File: src/Pcl.NET/VectorFloat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace Pcl.NET
{
    public class VectorFloat : Vector<float>
    {
        public VectorFloat()
        {
            _ptr = Invoke.std_vector_float_ctor();
        }

        internal VectorFloat(IntPtr ptr)
        {
            _ptr = ptr;
            _suppressDispose = true;
        }

        public VectorFloat(long count)
        {
            _ptr = Invoke.std_vector_float_ctor_count((ulong)count);
        }

        public unsafe VectorFloat(float[] array)
        {
            _ptr = Invoke.std_vector_float_ctor_count((ulong)array.Length);
            CopyFromArray(array, DataU);
        }

        public override long Count => (long)Invoke.std_vector_float_size(_ptr);

        public override IntPtr Data => Invoke.std_vector_float_data(_ptr);

        public override void Add(float item)
        {
            ThrowIfDisposed();
            Invoke.std_vector_float_add(_ptr, item);
        }

        public override void At(long index, ref float value)
        {
            ThrowIfDisposed();
            Invoke.std_vector_float_at(_ptr, (ulong)index, ref value);
        }

        public override void Clear()
        {
            ThrowIfDisposed();
            Invoke.std_vector_float_clear(_ptr);
        }

        public override void Insert(long index, float item)
        {
            ThrowIfDisposed();
            Invoke.std_vector_float_insert(_ptr, (ulong)index, item);
        }

        public override void Resize(long size)
        {
            ThrowIfDisposed();
            Invoke.std_vector_float_resize(_ptr, (ulong)size);
        }

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.std_vector_float_delete(ref _ptr);
            }
        }
    }
}
